### PR TITLE
Move Task queuing to Taskhubs

### LINKED ISSUE #78: Restructure `TaskQueue` system to use `TaskHubs`
As part of #23 we have decided to redesign the `TaskQueue` linked-list system to use a probabalistic  `TaskHub` "wheel and spoke " style setup.

This issue is intended as a subissue of #23 designed to break the resultant work into manageable chunks.  

As per #23 the task to be executed next will be selected probabilistically based on its `weight`.

This MAY be semi-orthogonal to the `priority` system (under discussion) which can allow for tasks to be given a global higher priority, however this is to be discussed.

### PR DESCRIPTION
**Pull request by hmacdope**
Fixes #78

Moves task queuing from a linked list with deterministic ordering to a hub-spoke model with stochastic pickup of tasks based on `weights`. 

@dotsdl one key question is that  currently the [:ACTIONS]  relationships are built only when the tasks are queued (makes sense), meaning weights cannot be set or interrogated before then. This is not explicitly checked for at this stage but are we happy with this limitation?

### CODE REVIEW

**Comment by hmacdope** [comment]
@dotsdl did you push the changes from our pair session ?

**Comment by dotsdl** [comment]
Apologies, I didn't. Just pushed now.

**Comment by dotsdl** [comment]
I see them on the branch: https://github.com/hmacdope/alchemiscale/tree/taskhub

**Comment by dotsdl** [comment]
Oh nevermind, it's further up because the commits were 18 hours ago. :facepalm:

File: alchemiscale/storage/statestore.py [hunk 1]
```diff
@@ -779,49 +826,138 @@ def queue_taskqueue_tasks(
...
+            WHERE NOT (th)-[:ACTIONS* {{taskhub: '{taskhub}'}}]->(tn)
               AND other_task.status = 'complete'
 
-            // create the connections that add it to the end of the queue, and delete old queue tail connection
-            CREATE (tqt)-[:FOLLOWS {{taskqueue: '{taskqueue}'}}] ->
-                      (tn)-[:FOLLOWS {{taskqueue: '{taskqueue}'}}]->(last)
-            DELETE tqtl
-
+            // create the connections 
+            CREATE (th)-[ar:ACTIONS {{taskhub: '{taskhub}'}}] ->
```
Comment (dotsdl) [inline alchemiscale/storage/statestore.py:847]: Might want to make this a MERGE statement instead of CREATE; should make the whole operation idempotent if run multiple times, avoiding duplicate ACTIONS relationships.

**Comment by dotsdl** [comment]
I can go ahead and fix the merge conflicts here, unless you'd prefer to?

**Comment by hmacdope** [comment]
Yep please go ahead. let me know if anything is unclear as we did a bit of renaming. 
After merge conflicts resolved I will go through and fix up the test suite and implement the changes we discussed. 

-  Add error for querying the weights when they are null or unable to be set because the `:ACTIONS` relationship is not created. 
- Add a weights interface to `action_taskhub_tasks`.

File: alchemiscale/storage/statestore.py (hunk 1 above)
Comment (dotsdl) [reply alchemiscale/storage/statestore.py:847]: Ah, nevermind. I see we do a WHERE up above which should be sufficient to avoid hitting this if the relationship already exists.

File: alchemiscale/storage/statestore.py [hunk 2]
```diff
@@ -831,33 +967,51 @@ def dequeue_taskqueue_tasks(
...
-
-            // create connection between behind and ahead nodes
-            CREATE (behind)-[newf:FOLLOWS {{taskqueue: '{taskqueue}'}}]->(ahead)
-
-            // delete connections between node to remove and behind, ahead nodes
-            DELETE behindf, aheadf
-
+            // get our task hub, as well as the task we want to remove
+            MATCH (th:TaskHub {{_scoped_key: '{taskhub}'}})-[:ACTIONS]->(task:Task {{_scoped_key: '{t}'}}),
+            DETACH DELETE task
```
Comment (dotsdl) [inline alchemiscale/storage/statestore.py:972]: We don't want to actually delete the `Task` here; `cancel_tasks` should be the opposite of `action_tasks`, so it should remove the `ACTIONS` relationship between the given `Task`s and the `TaskHub`.

**Comment by dotsdl** [comment]
Okay, think I've got the merge conflicts sorted. Seeing test failures in the claiming method in statestore. Can you address?

**Comment by hmacdope** [comment]
Yep for sure.

**Comment by hmacdope** [comment]
@dotsdl My tests are hanging locally, I'll investigate

File: alchemiscale/storage/statestore.py (hunk 2 above)
Comment (hmacdope) [reply alchemiscale/storage/statestore.py:1009]: Sorted.

File: alchemiscale/tests/integration/compute/client/test_compute_client.py [hunk 3]
```diff
@@ -52,44 +52,49 @@ def test_api_check(
...
-
-        # claim two more tasks
-        task_sks2 = compute_client.claim_taskqueue_tasks(
-            taskqueue_sks[0], count=2, claimant="me"
-        )
-
-        assert task_sks2 == all_task_sks[1:]
+        assert task_sks[0] in all_task_sks
+
+        # TODO: Currently we only allow tasks to be added to a taskhub with an EXTENDS
```
Comment (hmacdope) [inline alchemiscale/tests/integration/compute/client/test_compute_client.py:84]: This had to be modified due to #84 blocking adding of `:EXTENDS` tasks using `action_tasks`. Will be fixed in a separate PR.

**Comment by hmacdope** [comment]
@dotsdl I think this is probably ready for another review :smiley:

File: alchemiscale/tests/integration/compute/client/test_compute_client.py (hunk 3 above)
Comment (dotsdl) [reply alchemiscale/tests/integration/compute/client/test_compute_client.py:84]: I added the first task in the chain for each of the three `transformations` in the `n4js_preloaded` fixture, so now we have multiple tasks to work with for this test.

File: alchemiscale/tests/integration/storage/test_statestore.py [hunk 4]
```diff
@@ -327,144 +327,201 @@ def test_get_tasks(self, n4js, network_tyk2, scope_test):
...
         network_sk2 = n4js.create_network(an2, scope_test)
-        taskqueue_sk2: ScopedKey = n4js.create_taskqueue(network_sk2)
+        taskhub_sk2: ScopedKey = n4js.create_taskhub(network_sk2)
 
-        task_sks_fail = n4js.action_tasks(task_sks, taskqueue_sk2)
+        task_sks_fail = n4js.action_tasks(task_sks, taskhub_sk2)
         assert all([i is None for i in task_sks_fail])
 
-    def test_cancel_task(self, n4js, network_tyk2, scope_test):
+    def test_action_task_nested_fails(self, n4js: Neo4jStore, network_tyk2, scope_test):
```
Comment (dotsdl) [inline alchemiscale/tests/integration/storage/test_statestore.py:447]: Not sure we really need this test; see comments on `action_tasks` check.

**Review by dotsdl — APPROVED** [review]
Great work @hmacdope! Some comments, but none blocking. Pushing my changes noted then proceeding to merge.

File: alchemiscale/storage/statestore.py [hunk 5]
```diff
@@ -919,14 +1052,43 @@ def get_taskqueue_tasks(
         else:
             return [ScopedKey.from_str(t["_scoped_key"]) for t in tasks]
 
-    def claim_taskqueue_tasks(
-        self, taskqueue: ScopedKey, claimant: str, count: int = 1
+    def get_taskhub_unclaimed_tasks(
```
[the remark this answers is not in the document]
Comment (dotsdl) [reply alchemiscale/storage/statestore.py:1074]: Ah, we do use it in tests; this is fine.

File: alchemiscale/tests/integration/storage/test_statestore.py (hunk 4 above)
Comment (hmacdope) [reply alchemiscale/tests/integration/storage/test_statestore.py]: Yep makes sense

File: alchemiscale/storage/statestore.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hmacdope) [reply alchemiscale/storage/statestore.py:1068]: I was anticipating it might be useful for tests as you saw, but also possibly if one wanted to query what tasks are still waiting on the TaskHub eg for checks on the total volume of work etc. Can always YAGNI it away if not needed.

File: alchemiscale/storage/statestore.py [hunk 6]
```diff
@@ -851,35 +905,114 @@ def action_tasks(
...
+                RETURN ar
+                """
+                with self.transaction() as tx:
+                    tx.run(q)
+
+        # TODO: add exception for nothing returned for matches, or some Tasks missing
+
+    def get_task_weights(
+        self, tasks: List[ScopedKey], taskhub: ScopedKey
+    ) -> Dict[ScopedKey, float]:
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply alchemiscale/storage/statestore.py]: :+1:

File: alchemiscale/storage/statestore.py [hunk 7]
```diff
@@ -851,35 +905,114 @@ def action_tasks(
...
+            for t in tasks:
+                q = f"""
+                MATCH (th:TaskHub {{_scoped_key: '{taskhub}'}})-[ar:ACTIONS]->(task:Task {{_scoped_key: '{t}'}})
+                SET ar.weight = {weight}
+                RETURN ar
+                """
+                with self.transaction() as tx:
+                    tx.run(q)
+
+        # TODO: add exception for nothing returned for matches, or some Tasks missing
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply alchemiscale/storage/statestore.py]: :+1:

File: alchemiscale/storage/statestore.py [hunk 8]
```diff
@@ -761,85 +820,80 @@ def get_taskqueue(
...
         If this Task has an EXTENDS relationship to another Task, that Task must
-        be 'complete' before this Task can be added to *any* TaskQueue.
+        be 'complete' before this Task can be added to *any* TaskHub.
 
         """
+        depth = lambda L: isinstance(L, list) and max(map(depth, L)) + 1
+        if depth(tasks) > 1:
+            raise ValueError(
+                "tasks must be a flat list of ScopedKeys, list with depth > 1 not allowed"
+            )
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply alchemiscale/storage/statestore.py]: Yeah I agree, too internal to need this kind of sanity checking.
